### PR TITLE
Use explicit unicode strings + break up salt.utils

### PR DESCRIPTION
**Pull request by terminalmage**
This PR is part of what will be an ongoing effort to use explicit unicode strings in Salt. Because Python 3 does not suport Python 2's raw unicode string syntax (i.e. `ur'\d+'`), we must use `salt.utils.locales.sdecode()` to ensure that the raw string is unicode.  However, because of how `salt/utils/__init__.py` has evolved into the hulking monstrosity it is today, this means importing a large module in places where it is not needed, which could negatively impact performance. For this reason, this PR also breaks out some of the functions from `salt/utils/__init__.py` into new/existing modules under `salt/utils/`. The long term goal will be that the modules within this directory do not depend on importing `salt.utils`.

A summary of the changes in this PR is as follows:

* Moves the following functions from `salt.utils` to new locations (including a deprecation warning if invoked from `salt.utils`): `to_bytes`, `to_str`, `to_unicode`, `str_to_num`, `is_quoted`, `dequote`, `is_hex`, `is_bin_str`, `rand_string`, `contains_whitespace`, `clean_kwargs`, `invalid_kwargs`, `which`, `which_bin`, `path_join`, `shlex_split`, `rand_str`, `is_windows`, `is_proxy`, `is_linux`, `is_darwin`, `is_sunos`, `is_smartos`, `is_smartos_globalzone`, `is_smartos_zone`, `is_freebsd`, `is_netbsd`, `is_openbsd`, `is_aix`
* Moves the functions already deprecated by @rallytime to the bottom of `salt/utils/__init__.py` for better organization, so we can keep the deprecated ones separate from the ones yet to be deprecated as we continue to break up `salt.utils`
* Updates `salt/*.py` and all files under `salt/client/` to use explicit unicode string literals
* Gets rid of implicit imports of `salt.utils` (e.g. `from salt.utils import foo` becomes `import salt.utils.foo as foo`)
* Renames the `test.rand_str` function to `test.random_hash` to more accurately reflect what it does
* Modifies `salt.utils.stringutils.random()` (née `salt.utils.rand_string()`) such that it returns a string matching the passed size. Previously this function would get `size` bytes from `os.urandom()`, base64-encode it, and return the result, which would in most cases not be equal to the passed size.

### CODE REVIEW

**Comment by damon-atkins** [comment]
FYI Re: https://github.com/saltstack/salt/pull/42433 `Only force saltenv/pillarenv to be a string when not None`
str() in PY2 will result in the variable not being Unicode consider
```
@six.python_2_unicode_compatible
six.u()
from __future__ import unicode_literals
decode()
```

**Comment by terminalmage** [comment]
This isn't the only PR that will be opened for this process, it's not a trivial matter to modify a couple thousand files. Part of the move to all-unicode strings will include replacing calls to `str()` with calls to `six.text_type()`, except in instances where Python 2 can't support the `unicode` type, like setting a class' `__name__` attribute:

```python
>>> class Foo(object):
...     pass
...
>>> Foo.__name__ = u'Foo'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can only assign string to Foo.__name__, not 'unicode'
```

**Comment by rallytime** [comment]
I was just about to run these again, but it looks like one VM came back with some test results:
https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/13210/#showFailuresLink

**Comment by terminalmage** [comment]
Go Go Jenkins!

**Comment by terminalmage** [comment]
Tests have been passing on this PR since Friday. Sorry for not updating this issue sooner.

**Comment by rallytime** [comment]
Awesome! I have this on my list to start reviewing first thing tomorrow.

**Comment by terminalmage** [comment]
You are a brave soul. :smile: 

dangerous

File: salt/minion.py [hunk 1]
```diff
@@ -315,17 +319,17 @@ def load_args_and_kwargs(func, args, data=None, ignore_invalid=False):
                     # list of positional arguments. This keyword argument is
                     # invalid.
                     for key, val in six.iteritems(string_kwarg):
-                        invalid_kwargs.append('{0}={1}'.format(key, val))
+                        invalid_kwargs.append(r'{0}={1}'.format(key, val))
```
Comment (rallytime) [inline salt/minion.py]: Why is this using an `r'` instead of a `u'` here?

**Review by rallytime — APPROVED** [review]
@terminalmage Well done! This is very thorough. I only had a couple of minor things that I noticed during my review.

Since this change is so large and subject to conflicts, I think we should get this in now and have the issues I mentioned above addressed in a follow-up PR.

**Comment by terminalmage** [comment]
Right on all counts. Will have a PR open shortly to fix those three things.

**Comment by terminalmage** [comment]
Opened https://github.com/saltstack/salt/pull/42831 to address the problems @rallytime caught.

**Pull request merged**